Do not require a default datasource definition if TenantConnectionResolver is implemented

### Description

Hi there,

Would it be possible / an okay idea to not require a default datasource definition if multi-tenancy / programmatic datasource configuration is used? I'm not sure how at all this would be implemented, but from a developer experience point-of-view, this would be nice. No having to define extra non-used properties in application.properties hanging around just to avoid errors when the application tries to start.

### Implementation ideas

Perhaps a sort of escape hatch, where if an implementation of TenantConnectionResolver is detected in the CDI container, a message is logged letting the developer know that they are now on their own for validating their configuration and that quarkus will not check to make sure a datasource is configured before starting.

Checking with Quarkus 3.24, we still have a problem, but it shows up differently:

```
2025-06-30 14:03:22,011 ERROR [io.qua.run.Application] (main) Failed to start application: java.lang.RuntimeException: Failed to start quarkus                                                                                                                          
        at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)                                                                
        at io.quarkus.runtime.Application.start(Application.java:101)                                                               
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:119)                                 
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:80)                                                                          
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:51)                                                                          
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:144)                                                                         
        at io.quarkus.runner.GeneratedMain.main(Unknown Source)                                                                     
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:68)                                           
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:36)                                            
Caused by: io.quarkus.runtime.configuration.ConfigurationException: Unable to find datasource '<default>' for persistence unit '<default>': Bean is not active: SYNTHETIC bean [class=io.agroal.api.AgroalDataSource, id=sqqLi56D50iCdXmOjyjPSAxbLu0]                   
Reason: Datasource '<default>' was deactivated automatically because its URL is not set. To activate the datasource, set configuration property 'quarkus.datasource.jdbc.url'. Refer to https://quarkus.io/guides/datasource for guidance.                              
To avoid this exception while keeping the bean inactive:                                                                            
        - Configure all extensions consuming this bean as inactive as well, if they allow it, e.g. 'quarkus.someextension.active=false'                                                                                                                                 
        - Make sure that custom code only accesses this bean if it is active                                                        
        at io.quarkus.hibernate.orm.runtime.PersistenceUnitUtil.unableToFindDataSource(PersistenceUnitUtil.java:115)                
        at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.injectDataSource(FastBootHibernatePersistenceProvider.java:412)                                                                                                                        
        at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.buildRuntimeSettings(FastBootHibernatePersistenceProvider.java:224)                                                                                                                    
        at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.getEntityManagerFactoryBuilderOrNull(FastBootHibernatePersistenceProvider.java:194)                                                                                                    
        at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.createEntityManagerFactory(FastBootHibernatePersistenceProvider.java:67)                                                                                                               
        at jakarta.persistence.Persistence.createEntityManagerFactory(Persistence.java:90)                                          
        at jakarta.persistence.Persistence.createEntityManagerFactory(Persistence.java:66)                                          
        at io.quarkus.hibernate.orm.runtime.JPAConfig$LazyPersistenceUnit.get(JPAConfig.java:163)                                   
        at io.quarkus.hibernate.orm.runtime.JPAConfig$1.run(JPAConfig.java:63)                                                      
        at java.base/java.lang.Thread.run(Thread.java:840)                                                                          
Caused by: io.quarkus.arc.InactiveBeanException: Bean is not active: SYNTHETIC bean [class=io.agroal.api.AgroalDataSource, id=sqqLi56D50iCdXmOjyjPSAxbLu0]                                                                                                              
Reason: Datasource '<default>' was deactivated automatically because its URL is not set. To activate the datasource, set configuration property 'quarkus.datasource.jdbc.url'. Refer to https://quarkus.io/guides/datasource for guidance.                              
To avoid this exception while keeping the bean inactive:                                                                            
        - Configure all extensions consuming this bean as inactive as well, if they allow it, e.g. 'quarkus.someextension.active=false'                                                                                                                                 
        - Make sure that custom code only accesses this bean if it is active                                                        
        at io.agroal.api.AgroalDataSource_sqqLi56D50iCdXmOjyjPSAxbLu0_Synthetic_Bean.doCreate(Unknown Source)                       
        at io.agroal.api.AgroalDataSource_sqqLi56D50iCdXmOjyjPSAxbLu0_Synthetic_Bean.create(Unknown Source)                         
        at io.agroal.api.AgroalDataSource_sqqLi56D50iCdXmOjyjPSAxbLu0_Synthetic_Bean.create(Unknown Source)                         
        at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:119)                           
        at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:38)                                           
        at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:35)                                           
        at io.quarkus.arc.generator.Default_jakarta_enterprise_context_ApplicationScoped_ContextInstances.c10(Unknown Source)       
        at io.quarkus.arc.generator.Default_jakarta_enterprise_context_ApplicationScoped_ContextInstances.computeIfAbsent(Unknown Source)                                                                                                                               
        at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:35)                                             
        at io.quarkus.arc.impl.ClientProxies.getApplicationScopedDelegate(ClientProxies.java:23)                                    
        at io.agroal.api.AgroalDataSource_sqqLi56D50iCdXmOjyjPSAxbLu0_Synthetic_ClientProxy.arc$delegate(Unknown Source)            
        at io.agroal.api.AgroalDataSource_sqqLi56D50iCdXmOjyjPSAxbLu0_Synthetic_ClientProxy.arc_contextualInstance(Unknown Source)  
        at io.quarkus.arc.ClientProxy.unwrap(ClientProxy.java:52)                                                                   
        at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.injectDataSource(FastBootHibernatePersistenceProvider.java:410)                                                                                                                        
        ... 8 more                                                                                                                  
```

--- a/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/FastBootHibernatePersistenceProvider.java
+++ b/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/FastBootHibernatePersistenceProvider.java
@@ -36,6 +36,7 @@ import io.quarkus.hibernate.orm.runtime.boot.registry.PreconfiguredServiceRegist
 import io.quarkus.hibernate.orm.runtime.config.DatabaseOrmCompatibilityVersion;
 import io.quarkus.hibernate.orm.runtime.integration.HibernateOrmIntegrationRuntimeDescriptor;
 import io.quarkus.hibernate.orm.runtime.integration.HibernateOrmIntegrationRuntimeInitListener;
+import io.quarkus.hibernate.orm.runtime.migration.MultiTenancyStrategy;
 import io.quarkus.hibernate.orm.runtime.recording.PrevalidatedQuarkusMetadata;
 import io.quarkus.hibernate.orm.runtime.recording.RecordedState;
 
@@ -222,7 +223,8 @@ public final class FastBootHibernatePersistenceProvider implements PersistencePr
         Optional<String> dataSourceName = recordedState.getBuildTimeSettings().getSource().getDataSource();
         if (dataSourceName.isPresent()) {
             // Inject the datasource
-            injectDataSource(persistenceUnitName, dataSourceName.get(), runtimeSettingsBuilder);
+            injectDataSource(persistenceUnitName, dataSourceName.get(), recordedState.getMultiTenancyStrategy(),
+                    runtimeSettingsBuilder);
         }
 
         // Inject runtime configuration if the persistence unit was defined by Quarkus configuration
@@ -397,8 +399,10 @@ public final class FastBootHibernatePersistenceProvider implements PersistencePr
     }
 
     private static void injectDataSource(String persistenceUnitName, String dataSourceName,
+            MultiTenancyStrategy multiTenancyStrategy,
             RuntimeSettings.Builder runtimeSettingsBuilder) {
-        // first convert
+        // Note: the counterpart of this code, but for multitenancy (injecting the connection provider),
+        // can be found in io.quarkus.hibernate.orm.runtime.boot.FastBootMetadataBuilder.mergeSettings
 
         if (runtimeSettingsBuilder.isConfigured(AvailableSettings.URL) ||
                 runtimeSettingsBuilder.isConfigured(AvailableSettings.DATASOURCE) ||
@@ -407,6 +411,16 @@ public final class FastBootHibernatePersistenceProvider implements PersistencePr
                 runtimeSettingsBuilder.isConfigured(AvailableSettings.JAKARTA_JTA_DATASOURCE) ||
                 runtimeSettingsBuilder.isConfigured(AvailableSettings.JAKARTA_NON_JTA_DATASOURCE)) {
             // the datasource has been defined in the persistence unit, we can bail out
+            return;
+        }
+
+        if (multiTenancyStrategy != null && multiTenancyStrategy != MultiTenancyStrategy.NONE
+                && multiTenancyStrategy != MultiTenancyStrategy.DISCRIMINATOR) {
+            // Nothing to do: the datasource retrieval will be handled dynamically
+            // by an implementation of TenantConnectionResolver -- for instance
+            // io.quarkus.hibernate.orm.runtime.tenant.DataSourceTenantConnectionResolver.
+            // In the case of database multi-tenancy, that connection resolver
+            // might not even use the assigned datasource.
             return;
         }
 

--- a/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/boot/FastBootMetadataBuilder.java
+++ b/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/boot/FastBootMetadataBuilder.java
@@ -248,6 +248,9 @@ public class FastBootMetadataBuilder {
 
         if (multiTenancyStrategy != null && multiTenancyStrategy != MultiTenancyStrategy.NONE
                 && multiTenancyStrategy != MultiTenancyStrategy.DISCRIMINATOR) {
+            // Note: the counterpart of this code, but for single-tenancy (injecting the datasource),
+            // can be found in io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.injectDataSource
+
             // We need to initialize the multi tenant connection provider
             // on static init as it is used in MetadataBuildingOptionsImpl
             // to determine if multi-tenancy is enabled.

--- a/integration-tests/hibernate-orm-tenancy/connection-resolver-legacy-qualifiers/src/main/resources/application.properties
+++ b/integration-tests/hibernate-orm-tenancy/connection-resolver-legacy-qualifiers/src/main/resources/application.properties
@@ -6,8 +6,10 @@ quarkus.hibernate-orm.multitenant=database
 quarkus.hibernate-orm.dialect=MariaDB
 quarkus.hibernate-orm.packages=io.quarkus.it.hibernate.multitenancy.fruit
 
-# We create datasources manually, so a lack of configuration doesn't mean Quarkus should step in with defaults.
-quarkus.datasource.devservices.enabled=false
+# We create datasources programmatically, so we don't need the default datasource.
+# This makes sure Quarkus won't create a default datasource in test/dev mode.
+# This also helps reproduce https://github.com/quarkusio/quarkus/issues/29269.
+quarkus.datasource.active=false
 
 # Inventory persistence unit
 quarkus.hibernate-orm."inventory".schema-management.strategy=none

--- a/integration-tests/hibernate-orm-tenancy/connection-resolver/src/main/resources/application.properties
+++ b/integration-tests/hibernate-orm-tenancy/connection-resolver/src/main/resources/application.properties
@@ -6,8 +6,10 @@ quarkus.hibernate-orm.multitenant=database
 quarkus.hibernate-orm.dialect=MariaDB
 quarkus.hibernate-orm.packages=io.quarkus.it.hibernate.multitenancy.fruit
 
-# We create datasources manually, so a lack of configuration doesn't mean Quarkus should step in with defaults.
-quarkus.datasource.devservices.enabled=false
+# We create datasources programmatically, so we don't need the default datasource.
+# This makes sure Quarkus won't create a default datasource in test/dev mode.
+# This also helps reproduce https://github.com/quarkusio/quarkus/issues/29269.
+quarkus.datasource.active=false
 
 # Inventory persistence unit
 quarkus.hibernate-orm."inventory".schema-management.strategy=none

